heron/packing/src/java/com/twitter/heron/packing/binpacking/FirstFitDecreasingPacking.java

```diff
@@ -231,36 +226,135 @@ public void close() {
+    for (Map.Entry<String, Integer> entry : componentChanges.entrySet()) {
+      if (entry.getValue() > 0) {
+        componentsToScaleUp.put(entry.getKey(), entry.getValue());
+      }
+    }
+    return componentsToScaleUp;
   }
 
   /**
-   * Place a set of instances into container using the FFD heuristic. Adds new containers and
-   * instances starting at {@code firstContainerIndex) and {@code firstTaskIndex}, respectively.
+   * Generates the containers that correspond to the current packing plan
+   * along with their associated instances.
    *
-   * @return true if a placement was found, false otherwise
+   * @return List of containers for the current packing plan
    */
-  private Map<Integer, List<InstanceId>> assignInstancesToContainers(
-      Map<String, Integer> parallelismMap, int firstContainerIndex, int firstTaskIndex) {
-    Map<Integer, List<InstanceId>> allocation = new HashMap<>();
+  private ArrayList<Container> getContainers(PackingPlan currentPackingPlan) {
     ArrayList<Container> containers = new ArrayList<>();
+    Object[] currentContainers = currentPackingPlan.getContainers().toArray();
+    //sort containers based on containerIds;
+    Arrays.sort(currentContainers);
+    for (int i = 0; i < currentContainers.length; i++) {
+      int containerId = allocateNewContainer(containers);
+      for (PackingPlan.InstancePlan instancePlan
+          : ((PackingPlan.ContainerPlan) currentContainers[i]).getInstances()) {
+        containers.get(containerId - 1).add(instancePlan.getResource(),
+            new InstanceId(instancePlan.getComponentName(), instancePlan.getTaskId(),
+                instancePlan.getComponentIndex()));
+      }
+    }
+    return containers;
+  }
+
+  /**
+   * Generates an instance allocation for the current packing plan
+   *
+   * @return Map &lt; containerId, list of InstanceId belonging to this container &gt;
+   */
+  private Map<Integer, List<InstanceId>> getAllocation(PackingPlan currentPackingPlan) {
+    Map<Integer, List<InstanceId>> allocation = new HashMap<Integer, List<InstanceId>>();
+    for (PackingPlan.ContainerPlan containerPlan : currentPackingPlan.getContainers()) {
+      ArrayList<InstanceId> instances = new ArrayList<InstanceId>();
+      for (PackingPlan.InstancePlan instance : containerPlan.getInstances()) {
+        instances.add(new InstanceId(instance.getComponentName(), instance.getTaskId(),
+            instance.getComponentIndex()));
+      }
+      allocation.put(containerPlan.getId(), instances);
+    }
+    return allocation;
+  }
+
+  /**
+   * Assigns instances to containers
+   * @param containers helper data structure that describes the containers' status
+   * @param allocation existing packing plan
+   * @param parallelismMap component parallelism
+   * @param firstTaskIndex first taskId to use for the new instances
+   */
+  private void assignInstancesToContainers(
+      ArrayList<Container> containers, Map<Integer, List<InstanceId>> allocation,
```

Comment:
I feel like we're stretching our use of Lists and Maps in the signature here, by passing a mutable list of containers and a map of containers to instance lists. Instead can we use an object that encapsulated all of that? Like a PackingPlan object or a variant of it? Some container object that is able to effectively mutate a plan?